Form: Add support for labels in email subject, fix form validation etc.

EreMaijala opened the pull request:
This started from two requirements: 

1.) Need to be able to use select field labels in email subject
2.) Need to hide select values from the form

The other refactoring and fixes revolve around these and are so intertwined that I couldn't split this into more sensible pieces. Some notes:

1.) Avoid translating stuff early. Leave it more to the UI layer. This makes the separation more well-defined and would make it possible to use the data in other ways than just an email message (like storing feedback in database).
2.) formatEmailMessage is deprecated. I don't think the form should return the template name to use.
3.) aria attributes were removed and changed to fix HTML validation.
4.) Element id's were modified to make it less likely for them to collide with other UI elements. Also, this would make it possible to display multiple forms on a page without duplicate id's.

EreMaijala committed:
Form: Add support for labels in email subject, fix form validation etc.

demiankatz commented:
@EreMaijala, I just merged #2224, which introduced some conflicts here. Can you please resolve them when time permits?

EreMaijala committed:
Merge commit 'd8f76012c52f17e31cdbbcd9fd700ceb5038a21e' into dev-9.0-feedback-form-refactor

EreMaijala committed:
Improve the feedback sent response page.

No longer displays the form when feedback has been sent outside of lightbox. Also fixes escaping of page title, though will require a separate fix to layout.phtml to avoid double-escaping.

EreMaijala commented:
@demiankatz That was to be expected. Now resolved.

EreMaijala commented:
I also added a new page to be displayed when the form has been sent. This avoids displaying the form when not in lightbox.

demiankatz commented on themes/root/templates/Email/form.phtml:
```diff
@@ -1,13 +1,12 @@
 -----
 <?php else: ?>
-<?= ($label ? "$label: " : '') . $data['value'] . PHP_EOL ?>
+<?= ($label ? "$label: " : '') . implode(', ', array_map([$this, 'translate'], (array)$data['value'])) . PHP_EOL ?>
```
Is there ever a value that you DON'T want to translate? Could this cause weird things to happen to user input, or is the idea that we're always going to use textareas for free-form user input?

EreMaijala replied:
Good point. Indeed there is, and this didn't make sense at all. I changed it to translate only values of checkbox, radio and select.

demiankatz reviewed (commented):
It's a bit hard to review this many changes at once, but I did have one question (see below). Additionally, a few other things:

1.) The Mink FeedbackTest is failing... I'm guessing this is because you changed some IDs, so it can't select the elements it's expecting. Let me know if you need help reproducing/fixing it.

2.) I spotted a couple of small changes here that look like bug fixes -- things like adding parentheses around a `!$x = $this->function()` condition in an if statement. Is it worth pulling some of those back to dev so we can benefit from them sooner, or are they not significant enough?

Bottom line: once tests are passing, I'm reasonably comfortable with merging this... but let me know if there are any specific areas you'd like me to review with particular care.

EreMaijala committed:
Only translate values of checkbox, radio and select fields.

EreMaijala committed:
Fix Mink test.

EreMaijala committed:
Make textarea fields look better in the email messages.

This makes e.g. consecutive textarea fields more readable.

EreMaijala commented:
@demiankatz Sorry about making this so big. Most of the changes alone don't make that much sense, which caused things to pile up..

1.) Fixed.

2.) I don't think the fixes are significant enough. The particular case you mentioned seems to work without the parenthesis as well, but I'd rather have them in to make the intention it more obvious and readable.

I think the most important aspect to test is that translations still work properly and the resulting email looks ok and contains the right stuff. Also the different placeholders in email subject since that was one of the main drivers for these changes.

The code could use some more tests. Mink tests could include a more complex form, and unit tests should e.g. verify the contents of the resulting email message. But those have to be left for another task. I recorded them in https://vufind.org/jira/browse/VUFIND-1524.

demiankatz commented:
I can't seem to break it, so I have merged it. Hopefully there will be time for improved testing soon, in case we missed anything.
